esp/esdllib/esdl_xpath_extensions_libxml.cpp

```diff
@@ -472,6 +474,326 @@ static void strTokenizeFunction(xmlXPathParserContextPtr ctxt, int nargs)
+                xmlXPathSetError((ctxt), XPATH_EXPR_ERROR);
+                return;
+            }
+        }
+        xmlFree(tmp);
+    }
+
+    StringBuffer output;
+    txSummary->serialize(output, level, group, style);
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar*)output.str()));
+}
+
+static void deprecatedEncryptString(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toEncrypt = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    StringBuffer encrypted;
+    try
+    {
+        encrypt(encrypted, (const char*)toEncrypt);
+    }
+    catch (IException* e)
+    {
+        e->Release();
+        xmlXPathSetError(ctxt, XPATH_INVALID_OPERAND);
+    }
+    catch (...)
+    {
+        xmlXPathSetError(ctxt, XPATH_INVALID_OPERAND);
+    }
+    if (!xmlXPathCheckError(ctxt))
+        xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)encrypted.str()));
+    xmlFree(toEncrypt);
+}
+
+static void deprecatedDecryptString(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toDecrypt = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    StringBuffer decrypted;
+    try
+    {
+        decrypt(decrypted, (const char*)toDecrypt);
+    }
+    catch (IException* e)
+    {
+        e->Release();
+        xmlXPathSetError(ctxt, XPATH_INVALID_OPERAND);
+    }
+    catch (...)
+    {
+        xmlXPathSetError(ctxt, XPATH_INVALID_OPERAND);
+    }
+    if (!xmlXPathCheckError(ctxt))
+        xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)decrypted.str()));
+    xmlFree(toDecrypt);
+}
+
+static void encodeBase64String(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (0 == nargs || nargs > 2)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toEncode = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+    bool lineBreaks = false;
+    if (2 == nargs)
+    {
+        lineBreaks = xmlXPathPopBoolean(ctxt);
+        if (xmlXPathCheckError(ctxt))
+        {
+            xmlFree(toEncode);
+            return;
+        }
+    }
+
+    StringBuffer encoded;
+    JBASE64_Encode((const char*)toEncode, long(strlen((const char*)toEncode)), encoded, lineBreaks);
+    xmlXPathReturnString(ctxt, xmlStrdup((const xmlChar *)encoded.str()));
+    xmlFree(toEncode);
+}
+
+static void decodeBase64String(xmlXPathParserContextPtr ctxt, int nargs)
+{
+    if (nargs != 1)
+    {
+        xmlXPathSetArityError(ctxt);
+        return;
+    }
+
+    xmlChar *toDecode = xmlXPathPopString(ctxt);
+    if (xmlXPathCheckError(ctxt)) //includes null check
+        return;
+
+    StringBuffer decoded;
+    JBASE64_Decode((const char*)toDecode, decoded);
```

Comment:
In both cases do you mean the data has to be text only?  You're making a huge assumption here that this isn't true binary data.  I'm not sure doing this as an xpath function is a good idea.  If we do this we need to understand why the script needs access to the binary data, and how they would actually manipulate it.  I could see for example sending an image to an http call decoded, but that wouldn't work through this mechanism.  Unless you have some good use cases for a pure text version, and can put some safeguards around this that it really will be text, base64 may be another thing we have to separate out of this PR.

Reply:
I am making that assumption. The return types are limited, with only one ("external") that could possibly contain true binary data. We have no support for that type.

The function name is also significant. The presence of "String" in the name can refer to both the input parameter type and the result type. We can't return true binary in a string without some sort of encoding or escaping of bytes.

If we want to (and can) support true binary, I would consider a new function like `decodeBase64Binary`. I'm not sure it's worth holding this out because somebody could try to decode binary data and gets garbled text instead.

Reply:
Right.  Thought about this some more and I'm sure scapps has cases where it base64s text content so this is fine.  We need use cases for binary before designing anything.